sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -0,0 +1,643 @@
+    }
+
+    public Read<T> withDataSourceProviderFn(
+        SerializableFunction<Void, DataSource> dataSourceProviderFn) {
+      return toBuilder().setDataSourceProviderFn(dataSourceProviderFn).build();
+    }
+
+    public Read<T> fromQuery(String query) {
+      return toBuilder().setQuery(query).build();
+    }
+
+    public Read<T> fromTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Read<T> withStagingBucketName(String stagingBucketName) {
+      return toBuilder().setStagingBucketName(stagingBucketName).build();
+    }
+
+    public Read<T> withIntegrationName(String integrationName) {
+      return toBuilder().setIntegrationName(integrationName).build();
+    }
+
+    public Read<T> withCsvMapper(CsvMapper<T> csvMapper) {
+      return toBuilder().setCsvMapper(csvMapper).build();
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
+      return toBuilder().setCoder(coder).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      // Either table or query is required. If query is present, it's being used, table is used
+      // otherwise
+      checkArgument(
+          getQuery() != null || getTable() != null, "fromTable() or fromQuery() is required");
+      checkArgument(
+          !(getQuery() != null && getTable() != null),
+          "fromTable() and fromQuery() is not allowed together");
+      checkArgument(getCsvMapper() != null, "withCsvMapper() is required");
+      checkArgument(getCoder() != null, "withCoder() is required");
+      checkArgument(getIntegrationName() != null, "withIntegrationName() is required");
+      checkArgument(getStagingBucketName() != null, "withStagingBucketName() is required");
+      checkArgument(
+          (getDataSourceProviderFn() != null),
+          "withDataSourceConfiguration() or withDataSourceProviderFn() is required");
+
+      String gcpTmpDirName = makeTmpDirName();
+      PCollection<Void> emptyCollection = input.apply(Create.of((Void) null));
+
+      PCollection<T> output =
+          emptyCollection
+              .apply(
+                  ParDo.of(
+                      new CopyIntoStageFn(
+                          getDataSourceProviderFn(),
+                          getQuery(),
+                          getTable(),
+                          getIntegrationName(),
+                          getStagingBucketName(),
+                          gcpTmpDirName,
+                          getSnowflakeService())))
+              .apply(FileIO.matchAll())
+              .apply(FileIO.readMatches())
+              .apply(readFiles())
+              .apply(ParDo.of(new MapCsvToStringArrayFn()))
+              .apply(ParDo.of(new MapStringArrayToUserDataFn<>(getCsvMapper())));
+
+      output.setCoder(getCoder());
+
+      emptyCollection
+          .apply(Wait.on(output))
+          .apply(
+              ParDo.of(
+                  new CleanTmpFilesFromGcsFn(
```

Comment:
It seems that it doesn't at the moment. @kkucharc will share more info here

Reply:
Yes, we checked and it doesn't. @aromanenko-dev do you think it should be provided? In case of testing, probably tests should take care of cleanup.

Reply:
I'm afraid that many failed pipelines could lead to wasting of used dick space in this case. It would be better to avoid such behavior, if possible.

Reply:
While I generally agree that failing pipelines will eventually lead to lots of garbage on GCS (or other storage in future) I'm not sure how to ensure those files are deleted regardless of pipeline status. 

AFAIK there is nothing like `@After` or `@Teardown` for `PTransform` which we have in our IO. Using `Wait` transform gives us ability to remove files once data is read and seems reasonable choice.

I've been checking how i.e. `BigQueryIO` is handling that case as it also needs to cleanup and they also have a cleanup transform that is called once all rows are read. I assume in their case cleanup also won't be run if something in-between fails.

How about filing JIRA issue for this case and try to look at solution in parallel to delivering other pieces of the Snowflake connector? @aromanenko-dev @kkucharc  WDYT?

Reply:
This is a known issue with BQ source as well. Failed pipelines can leave temporary files behind. I'm afraid there is no good solution today. I think we need to introduce some sort of a generalized cleanup step to address this.

Reply:
In this case, I'd suggest to add this into IO class Java doc to make users aware that such situation is possible and it will require manual procedure to clean temp dirs up.

Reply:
I've added a note [here](https://github.com/apache/beam/pull/11360/commits/6f3ff11d758a45619ff25b6e5f270c2c970fce05#diff-de3ace437cd24772382c19f0c5d9a062R115) - would it be enough?